### PR TITLE
Add a GUC to display create gang time while executing statements

### PR DESCRIPTION
**Pull request by zxuejing**
Add GUC gp_print_create_gang_time control whether to print information about creating gang time.
We print the create gang time for both DDL and DML.

If all the segDescs of a gang are from the cached pool, we regard the gang as reused.
We only display the shortest and longest establish conn time and their segindexs of a gang.

The info of the shortest establish conn time and the longest establish conn time is the same for 1-gang.

DDL:
```
create table t(tc1 int);
INFO: The shortest establish conn time: 4.48 ms, segindex: 2,
      The longest  establish  conn time: 8.13 ms, segindex: 1

set optimizer=off;
INFO: (Gang) is reused
```
DML:
we can use DML or explain analyze to get create gang time.
```
select * from t_create_gang_time t1, t_create_gang_time t2 where t1.tc1=2;
INFO: (Slice1) is reused
INFO: (Slice2) The shortest establish conn time: 4.80 ms, segindex: 0,
               The longest  establish conn time: 4.80 ms, segindex: 0
tc1 | tc2 | tc1 | tc2
-----+-----+-----+-----
(0 rows)

explain analyze select * from t_create_gang_time t1, t_create_gang_time t2 where t1.tc1=2;
INFO: (Slice1) is reused
INFO: (Slice2) is reused
QUERY PLAN
......
```

### CODE REVIEW

File: src/backend/cdb/dispatcher/cdbgang.c [hunk 1]
```diff
@@ -1136,3 +1136,63 @@ gp_backend_info(PG_FUNCTION_ARGS)
 	}
 #undef BACKENDINFO_NATTR
 }
+
+/*
+ * Print the time of create a gang.
+ * if all segDescs of the gang are cached, we regard the gang as reused.
+ * else we print the shortest time and the longest time of estabishing connection to the segDesc.
+ */
+void printCreateGangTime(int sliceId, Gang *gang)
```
Comment (Aegeaner) [inline src/backend/cdb/dispatcher/cdbgang.c:1145]: Format: should add a newline after the return type:
```
void
printCreateGangTime(int sliceId, Gang *gang)
```

File: src/include/utils/unsync_guc_name.h [hunk 2]
```diff
@@ -530,6 +530,7 @@
 		"temp_file_limit",
 		"test_AppendOnlyHash_eviction_vs_just_marking_not_inuse",
 		"test_print_direct_dispatch_info",
+		"debug_print_create_gang_time",
```
Comment (adam8157) [inline src/include/utils/unsync_guc_name.h:533]: Items in this file should be ordered.

File: src/backend/utils/misc/guc_gp.c [hunk 3]
```diff
@@ -1411,6 +1412,17 @@ struct config_bool ConfigureNamesBool_gp[] =
 		NULL, NULL, NULL
 	},
 
+	{
+		{"debug_print_create_gang_time", PGC_SUSET, DEVELOPER_OPTIONS,
+			gettext_noop("For testing purposes, print information about create gang time."),
+			NULL,
+			GUC_SUPERUSER_ONLY | GUC_NOT_IN_SAMPLE
```
Comment (adam8157) [inline src/backend/utils/misc/guc_gp.c:1419]: Has it to be super user only? I'm not sure.

File: src/backend/cdb/dispatcher/cdbgang_async.c [hunk 4]
```diff
@@ -139,6 +140,8 @@ cdbgang_createGang_async(List *segments, SegmentType segmentType)
 			if (segdbDesc->conn != NULL && !cdbconn_isBadConnection(segdbDesc))
 			{
 				connStatusDone[i] = true;
+				/* -1 means this connection is cached */
+				segdbDesc->establishConnTime = -1;
```
Comment (adam8157) [inline src/backend/cdb/dispatcher/cdbgang_async.c:144]: I'm fine with this, just a question: it should already be set by `cdbcomponent_recycleIdleQE()`, right?

File: src/backend/cdb/dispatcher/cdbgang_async.c [hunk 5]
```diff
@@ -14,6 +14,7 @@
  */
 
 #include "postgres.h"
+#include "utils/elog.h"
```
Comment (adam8157) [inline src/backend/cdb/dispatcher/cdbgang_async.c:17]: Is this needed? Or is `instr_time.h` good enough?

File: src/include/utils/unsync_guc_name.h (hunk 2 above)
Comment (zxuejing) [reply src/include/utils/unsync_guc_name.h:533]: ok, I will modify it.

File: src/backend/cdb/dispatcher/cdbgang_async.c (hunk 4 above)
Comment (zxuejing) [reply src/backend/cdb/dispatcher/cdbgang_async.c:144]: Yes, you are right,  segdbDesc->establishConnTime is set to -1 when cdbcomponent_recycleIdleQE. We can keep just one.

File: src/backend/cdb/dispatcher/cdbgang.c (hunk 1 above)
Comment (zxuejing) [reply src/backend/cdb/dispatcher/cdbgang.c:1145]: Thanks a lot, I will add a line break after the return type.

**Comment by fanfuxiaoran** [comment]
I can understand that printing time costs for a gang.  But for each slice, I just wonder in which situation we need to know the time cost by each slice to create a gang.

**Comment by zxuejing** [comment]
> I can understand that printing time costs for a gang. But for each slice, I just wonder in which situation we need to know the time cost by each slice to create a gang.

There is a Jira, some queries execute very slowly, we thought the plan is not good, But, after a lot of analysis, we found creating gang cost a lot of time through the log, so this can help to locate the real cause when some queries have poor performance.

File: src/backend/utils/misc/guc_gp.c [hunk 6]
```diff
@@ -1411,6 +1412,17 @@ struct config_bool ConfigureNamesBool_gp[] =
 		NULL, NULL, NULL
 	},
 
+	{
+		{"debug_print_create_gang_time", PGC_USERSET, DEVELOPER_OPTIONS,
```
[the remark this answers is not in the document]
Comment (zxuejing) [reply src/backend/utils/misc/guc_gp.c:1416]: Thanks, I will modify this, if this info is embedded in Explain analyze plans, DDL will not display the create gang time, and we also need to add a GUC to control DDL.

**Review by Aegeaner — APPROVED** [review]

**Review by adam8157 — APPROVED** [review]

**Pull request merged**